devices/flash-gr716/flashdrv.c

```diff
@@ -0,0 +1,377 @@
+	if ((offs < memsz) && ((offs + len) <= memsz)) {
+		return 1;
+	}
+
+	return 0;
+}
+
+
+static ssize_t fdrv_directSectorWrite(struct nor_device *dev, addr_t offs, const u8 *src)
+{
+	addr_t pos;
+	int res = nor_eraseSector(&dev->spimctrl, offs, dev->nor->tSE);
+	if (res < 0) {
+		return res;
+	}
+
+	for (pos = 0; pos < dev->nor->sectorSz; pos += dev->nor->pageSz) {
+		res = nor_pageProgram(&dev->spimctrl, offs + pos, src + pos, dev->nor->pageSz, dev->nor->tPP);
+		if (res < 0) {
+			return res;
+		}
+	}
+
+	return pos;
+}
+
+
+/* Device driver interface */
+
+
+static ssize_t flashdrv_read(unsigned int minor, addr_t offs, void *buff, size_t len, time_t timeout)
+{
+	int res;
+	struct nor_device *dev = fdrv_minorToDevice(minor);
+	size_t doneBytes = 0;
+
+	if (dev == NULL) {
+		return -ENXIO;
+	}
+
+	if (fdrv_isValidAddress(dev->nor->totalSz, offs, len) == 0) {
+		return -EINVAL;
+	}
+
+	if (len == 0u) {
+		return 0;
+	}
+
+	if (fdrv_getSectorAddress(dev, offs) == dev->sectorBufAddr) {
+		doneBytes = min(dev->sectorBufAddr + dev->nor->sectorSz - offs, len);
+
+		hal_memcpy(buff, dev->sectorBuf + offs - dev->sectorBufAddr, doneBytes);
+
+		if (len == doneBytes) {
+			return doneBytes;
+		}
+
+		offs += doneBytes;
+		buff += doneBytes;
+		len -= doneBytes;
+	}
+
+	res = nor_readData(&dev->spimctrl, offs, buff, len);
+
+	return (res < 0) ? res : (doneBytes + res);
+}
+
+
+static int flashdrv_sync(unsigned int minor)
+{
+	struct nor_device *dev = fdrv_minorToDevice(minor);
+
+	if (dev == NULL) {
+		return -ENXIO;
+	}
+
+	if ((dev->sectorBufAddr == (addr_t)-1) || (dev->sectorBufDirty == 0)) {
+		return EOK;
+	}
+
+	return fdrv_directSectorWrite(dev, dev->sectorBufAddr, dev->sectorBuf);
+}
+
+
+static ssize_t flashdrv_write(unsigned int minor, addr_t offs, const void *buff, size_t len)
+{
+	ssize_t res;
+	addr_t sectorOfs, lastSectorAddr, currAddr;
+	const u8 *src = buff;
+	size_t doneBytes = 0, chunk;
+	struct nor_device *dev = fdrv_minorToDevice(minor);
+
+	if (dev == NULL) {
+		return -ENXIO;
+	}
+
+	if (fdrv_isValidAddress(dev->nor->totalSz, offs, len) == 0) {
+		return -EINVAL;
+	}
+
+	if (len == 0u) {
+		return 0;
+	}
+
+	lastSectorAddr = fdrv_getSectorAddress(dev, offs + len - 1u);
+
+	while (doneBytes < len) {
+		currAddr = fdrv_getSectorAddress(dev, offs);
+
+		if ((currAddr != dev->sectorBufAddr) && (currAddr == lastSectorAddr)) {
+			/* Last sector to write to, cache it */
+			res = flashdrv_sync(minor);
+			if (res < 0) {
+				return res;
+			}
+
+			dev->sectorBufAddr = currAddr;
+			res = nor_readData(&dev->spimctrl, dev->sectorBufAddr, dev->sectorBuf, dev->nor->sectorSz);
+			if (res < 0) {
+				return res;
+			}
+		}
+
+		if (currAddr == dev->sectorBufAddr) {
+			/* Sector to write to in cache */
+			sectorOfs = offs - dev->sectorBufAddr;
+			chunk = min(dev->nor->sectorSz - sectorOfs, len - doneBytes);
+
+			hal_memcpy(dev->sectorBuf + sectorOfs, src, chunk);
+			dev->sectorBufDirty = 1;
+		}
+		else {
+			chunk = dev->nor->sectorSz;
```

Comment:
This case will be invoked when:
- currAddr is not cached,
- this is a first write,
- beginning is not aligned to the start of the sector.
Above condition will result in data loss (from beginning of the sector to the starting offset)